### ISSUE TITLE
Relax extractors types to allow non strings

### ISSUE THREAD

**Issue by danielepolencic** [FullHuman/purgecss#402]
Consider the following snippet of code that purges CSS given an HTML string:

```ts
purgecss({
      content: [{ raw: `<HTML string here>`, extension: 'html' }],
      extractors: [
        {
          extractor: (content: string) => [],
          extensions: ['html'],
        },
      ],
    })
```

the content of `raw` is passed as an argument to the `extractor` function.

The type of `raw` **must be** string.

```ts
interface RawContent {
  extension: string;
  raw: string;
}
```

This makes sense if the content is already a string. However, there are cases where this is not true.
As an example, I usually keep the representation of the DOM as an AST (abstract syntax tree) using [Hast](https://github.com/syntax-tree/hast).
Ideally, I'd like to pass the Hast tree in `raw` rather than serialise the AST to only later re-parse it.

I'd like to propose the following change: `raw` should not be restricted to a string. 
However, I don't have enough experience with the project to tell if there are further ramifications.

```ts
interface RawContent<T=string> {
  extension: string;
  raw: T;
}
```

the extractor could have a signature of `(content: T) => string[]`.

```ts
// previous
type ExtractorFunction = (content: string) => string[];
// proposed
type ExtractorFunction<T = string> = (content: T) => string[];
```

Proof of concept with code changes in #403

**Comment by Ffloriel** [comment]
Merged the PR. I think it's a good idea to keep the AST, I'd like to take a look at how you are currently using it if you have a repo available. Thanks

**Comment by danielepolencic** [comment]
https://github.com/learnk8s/learnk8s.io/blob/master/src/optimise.tsx#L209

**Comment by Ffloriel** [comment]
released in 2.3.0
